### ISSUE TITLE
line not ragged before euouae block in some cases

### ISSUE THREAD

**Issue by eroux** [gregorio-project/gregorio#984]
Compiling

```tex
\documentclass[12pt, a5paper]{book}
\usepackage{libertine}
\usepackage[autocompile]{gregoriotex}
\begin{document}
\gresetbarspacing{new}
\gresetbreakbeforeeuouae{ragged}
\gresetlastline{ragged}
\gregorioscore{a--prophetae-predicaverunt-MWE3}
\end{document}
```

with gabc

```
name: Prophetae predicaverunt;
office-part: antiphona;
mode:1;
%%
(c4) Pro(d)phé(ddo>)tæ(c) pre(d)di(f)ca(e)vé(d)runt(dc~) *() na(f)sci(g) De(h)us(g) Sal(h>)va(g)tó(f_0fg)rem(f_) (,) de(f) Vír(f)gi(g)ne(e) Ma(f)rí(d.)a.(d.) (::) <eu>E(h) u(h) o(g) u(f) a(gh) e</eu>(ggo>) (::)
```

one gets

![bug2](https://cloud.githubusercontent.com/assets/60868/13425160/60d5fa8a-dfa5-11e5-9dc9-5352d59a0752.png)

where the second line is stretched quite a lot, comparing to a version with a `Z`:

![bug3](https://cloud.githubusercontent.com/assets/60868/13425219/b025c732-dfa5-11e5-8328-1190a7d01431.png)

**Comment by henryso** [comment]
This is probably the (infamous) TeX linebreaking algorithm at work.  Notice that TeX even pulls from the first line in an attempt to justify the paragraph.  If that's the case, I don't know if there's much we can do besides noting the option to use `Z` when such is unacceptable.

**Comment by eroux** [comment]
I don't think there's such a thing as a bug in TeX, the problem here is in gregorio. If you try this patch on the example:

```diff
diff --git a/tex/gregoriotex-main.tex b/tex/gregoriotex-main.tex
index 3d273e6..be1b7ac 100644
--- a/tex/gregoriotex-main.tex
+++ b/tex/gregoriotex-main.tex
@@ -1643,6 +1643,7 @@
   \ifgre@raggedbreakbeforeeuouae %
     \directlua{gregoriotex.save_pos(#1,2)}%
     \ifcase\directlua{gregoriotex.is_ypos_different(#1)}\relax%
+      \directlua{texio.write_nl('debug2')}%
       \gre@newlinecommon{1}%
     \fi %
   \fi %
```

you'll see that the "debug2" is printed during the first compilation (which is a bit surprising), and then not printed anymore. This is what causes the problem. Can you take a look?

**Comment by henryso** [comment]
That behavior is not entirely surprising to me, but I don't deny that the behavior is buggy.  However, after looking at this for a while, I don't think I will be able to fix it in time for 4.1 (today).  In fact, I'm not even sure how to change the approach to make this work properly.

Currently, the code puts a macro on the syllable prior to one that starts the `<eu>` block after all glyphs but before any forced newline.  The y-position of this macro is compared with the y-position of the code listed above and, if they differ, `\gre@newlinecommon{1}` is put in to make the break ragged.

What I *think* is happening is this: In the case where the glues would not permit justification, these y-positions are different and it works ragged as expected.  In the case where the glues do permit justification, these y-positions are the same, and it ends up the way you see above.

This is likely to change with #985, so perhaps the best way is to note this behavior, having `Z` as the workaround until we can either fix this simultaneously with #985, unless, of course, someone can give me some hints as to how I might approach this.

Note: I tried putting the first macro before the last glyph or bar prior to the `<eu>` block, but that results in an unstable situation where the PDF never renders without the positions changing.

**Comment by eroux** [comment]
There are some things that really make no sense to me in gregoriotex.lua. For instance, apply this patch:

```diff
diff --git a/tex/gregoriotex.lua b/tex/gregoriotex.lua
index 8528ee3..df661d0 100644
--- a/tex/gregoriotex.lua
+++ b/tex/gregoriotex.lua
@@ -146,12 +146,18 @@ local function heights_changed()
   for id, tab in pairs(var_brace_positions) do
     if keys_changed(tab, new_var_brace_positions[id]) then return true end
   end
+  texio.write_nl('this is new pos:')
+  texio.write_nl(table.serialize(new_pos_saves))
+  texio.write_nl('this is old ones:')
+  texio.write_nl(table.serialize(pos_saves))
   for id, tab in pairs(new_pos_saves) do
     if keys_changed(tab, pos_saves[id]) then return true end
   end
   for id, tab in pairs(pos_saves) do
     if keys_changed(tab, new_pos_saves[id]) then return true end
   end
+  texio.write_nl('and yet I consider that nothing changed!')
+  texio.write_nl('')
   return false
 end
```

and run the example of this issue... I really don't see the point of not taking the new values into account...

**Comment by henryso** [comment]
It should certainly be taking the new values into account, so if it's not doing that, there's a bug somewhere.

**Comment by henryso** [comment]
Maybe something shifted from caring about just the keys to caring about keys and values.  I'm reviewing the history.

**Comment by eroux** [comment]
In fact it quite makes sense, because we just want the difference in the first run... I think this patch fixes the issue:

```diff
diff --git a/tex/gregoriotex-main.tex b/tex/gregoriotex-main.tex
index 3d273e6..c20784e 100644
--- a/tex/gregoriotex-main.tex
+++ b/tex/gregoriotex-main.tex
@@ -1640,15 +1640,15 @@
 \gre@in@euouaefalse
 
 \def\GreBeginEUOUAE#1{%
+  \ifgre@euouae@implies@nlba %
+    \GreBeginNLBArea{0}{0}%
+  \fi %
   \ifgre@raggedbreakbeforeeuouae %
     \directlua{gregoriotex.save_pos(#1,2)}%
     \ifcase\directlua{gregoriotex.is_ypos_different(#1)}\relax%
       \gre@newlinecommon{1}%
     \fi %
   \fi %
-  \ifgre@euouae@implies@nlba %
-    \GreBeginNLBArea{0}{0}%
-  \fi %
   \gre@in@euouaetrue %
   % as soon as the EUOUAE starts, we stop blocking the EOL custos that might
   % have been blocked on the prior syllable.
diff --git a/tex/gregoriotex.lua b/tex/gregoriotex.lua
index 8528ee3..e8b21b7 100644
--- a/tex/gregoriotex.lua
+++ b/tex/gregoriotex.lua
@@ -915,7 +915,7 @@ local function is_ypos_different(index)
       tex.sprint([[\number0\relax ]])
     end
   else
-    tex.sprint([[\number0\relax ]])
+    tex.sprint([[\number1\relax ]])
   end
 end
```

can you please try?

**Comment by henryso** [comment]
Are you saying that checking only that the tables have the same structure, regardless of the values, makes sense?  I'm confused if that's the case.

**Comment by henryso** [comment]
I actually did try that change in gregoriotex-main.tex (moving the NLBA part before the ragged check), but it didn't do anything.  However, perhaps the change to gregoriotex.lua fixes it.

**Comment by eroux** [comment]
about the check of the table structure, what it is basically if the previous had saved pos or not, and if it didn't it adds the new pos, but doesn't if there were already some pos. The code is quite fragile though...

**Comment by henryso** [comment]
I agree, it's a bug waiting to happen and should be fixed for 5.0.  I think your suggested fix works for this well enough to hold us over until #985 can be fixed and this fragility corrected.  I will push the change shortly once I have reviewed all the tests.

**Comment by eroux** [comment]
good, thanks!

**Comment by eroux** [comment]
Also, it would be important to fake a `\GreLastOfLine` for the syllable in which this happens... I don't think it's a lot of work, and it would make the output nicer

**Comment by henryso** [comment]
I can do the `\GreLastOfLine` thing (I think it involves the C code and a new macro), but I don't think I can get that done until much later tonight, so it's going to miss the deadline.

**Comment by eroux** [comment]
Ok, let's leave it for 5.0 then

**Comment by henryso** [comment]
I'll close this bug then as 4.1 and put in a new issue for `\GreLastOfLine` for 5.0.

**Comment by eroux** [comment]
There is still a problem with:

```tex
\documentclass[12pt, a5paper]{book}
\usepackage{libertine}
\usepackage[autocompile]{gregoriotex}
\begin{document}
\gresetbarspacing{new}
\gresetbreakbeforeeuouae{ragged}
\gresetlastline{ragged}
\gregorioscore{a--prophetae-predicaverunt-MWE3}
\end{document}
```

```
name: Prophetae predicaverunt;
mode:1;
%%
(c4) Pro(d)phé(ddo>)tæ(c) pre(d)di(f)cant(dc~) *() <v><<</v>ve(f)ni(f)et(g) Ie(h)sum!<v>>></v>(g) (,) de(f) Vír(f)gi(g)ne(e) Ma(f)rí(d.)a,(d.) ad(f) Vír(f)gi(g)nem(e) Ma(f)rí(d.)am...(d.) (;) de(f) Vír(f)gi(g)ne(e) Ma(f)rí(d.)a,(d.) ad(f) Vír(f)gi(g)nem(e) Ma(f)rí(d.)am...(d.) (,) de(f) Vír(f)gi(g)ne(e) Ma(f)rí(d.)a.(d.) (::) <eu>E(h) u(h) o(g) u(f) a(gh) e</eu>(g) (::)
```

giving

![bug](https://cloud.githubusercontent.com/assets/60868/13457205/397974f4-e067-11e5-8bc8-db5206c00add.png)

**Comment by henryso** [comment]
Removing from 4.1.

**Comment by eroux** [comment]
(reopening as it has only partially been fixed)
